test/e2e/azure_securitygroups.go

```diff
@@ -0,0 +1,188 @@
+			Destination:      pointer.String("*"),
+		}
+		testSecurityRule2 = infrav1.SecurityRule{
+			Name:             "test-security-rule-2",
+			Description:      "test-security-rule-2",
+			Protocol:         "Tcp",
+			Direction:        "Inbound",
+			Priority:         110,
+			SourcePorts:      pointer.String("*"),
+			DestinationPorts: pointer.String("80"),
+			Source:           pointer.String("*"),
+			Destination:      pointer.String("*"),
+		}
+		input AzureSecurityGroupsSpecInput
+	)
+
+	Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+
+	input = inputGetter()
+	Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+	Expect(input.Namespace).ToNot(BeNil(), "Invalid argument. input.Namespace can't be nil when calling %s spec", specName)
+	Expect(input.ClusterName).ToNot(BeEmpty(), "Invalid argument. input.ClusterName can't be empty when calling %s spec", specName)
+
+	By("creating a Kubernetes client to the workload cluster")
+	workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(ctx, input.Namespace.Name, input.ClusterName)
+	Expect(workloadClusterProxy).NotTo(BeNil())
+	mgmtClient := bootstrapClusterProxy.GetClient()
+	Expect(mgmtClient).NotTo(BeNil())
+
+	// get subscription id
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := azureutil.GetAuthorizer(settings)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("creating a subnets client")
+	subnetsClient := network.NewSubnetsClient(subscriptionID)
+	subnetsClient.Authorizer = auth
+
+	securityGroupsClient := network.NewSecurityGroupsClient(subscriptionID)
+	securityGroupsClient.Authorizer = auth
+
+	securityRulesClient := network.NewSecurityRulesClient(subscriptionID)
+	securityRulesClient.Authorizer = auth
+
+	azureCluster := &infrav1.AzureCluster{}
+	err = mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.InfrastructureRef.Namespace,
+		Name:      input.Cluster.Spec.InfrastructureRef.Name,
+	}, azureCluster)
+	Expect(err).NotTo(HaveOccurred())
+
+	var expectedSubnets infrav1.Subnets
+	checkSubnets := func(g Gomega) {
+		for _, expectedSubnet := range expectedSubnets {
+			securityGroup, err := securityGroupsClient.Get(ctx, azureCluster.Spec.ResourceGroup, expectedSubnet.SecurityGroup.Name, "")
+			g.Expect(err).NotTo(HaveOccurred())
+
+			securityRules, err := securityRulesClient.List(ctx, azureCluster.Spec.ResourceGroup, *securityGroup.Name)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			var expectedSecurityRuleNames []string
+			for _, expectedSecurityRule := range expectedSubnet.SecurityGroup.SecurityRules {
+				expectedSecurityRuleNames = append(expectedSecurityRuleNames, expectedSecurityRule.Name)
+			}
+
+			for _, securityRule := range securityRules.Values() {
+				g.Expect(expectedSecurityRuleNames).To(ContainElement(*securityRule.Name))
```

Comment:
Will this function catch when a security that should _not_ exist does exist? If I'm reading this right, it will check to see that all the expected ones are present, but doesn't explicitly check for the absence of ones which should not exist.

Reply:
nvm I might've been reading this line a bit backwards.

Reply:
Hmm I guess this doesn't check the case where if there is an expected security rule that has not yet been created on Azure, although it is a bit hard to compare the values of each list without two separate passes in the opposite direction. Do you think it's worth it to check the inverse here? There's probably a better way to do this as well.

Reply:
The AKS tags test needs to handle this kind of case where not all tags in Azure are expected to exist in the CAPZ spec. It does a first pass to figure out which ones are "extra," removes those from the "actual" tags when we need to compare, then does a direct equality check between the spec and modified "actual" tags.

Overall though I'd vote to do that as a follow-up to make sure we can get this fix in for the 1.10 release. This test as-is verifies the bug which is the most important thing.